shared/js/advanced_telemetry_helper.js

```diff
@@ -0,0 +1,154 @@
+ *  // are not required on subsequent calls.
+ *  ath.exp('mylinearmetric', 8000);
+ *
+ */
+
+/* globals SettingsListener */
+(function(exports) {
+  'use strict';
+
+  const APP_TELEMETRY_LOG_PREFIX = 'telemetry';
+  const ATH = AdvancedTelemetryHelper;
+  // This is exposed for unit testing purposes
+  ATH.TELEMETRY_ENABLED = false;
+  ATH.TELEMETRY_ENABLED_KEY = 'debug.performance_data.advanced_telemetry';
+  ATH.SETTINGS_LISTENER = false;
+
+  function AdvancedTelemetryHelper() {
+    AdvancedTelemetryHelper.prototype.getTelemetrySetting();
+  }
+
+  AdvancedTelemetryHelper.prototype.getTelemetrySetting = function() {
+    if (ATH.SETTINGS_LISTENER) {
+      return;
+    }
+
+    this.telemetryEnabledListener = function telemetryEnabledListener(enabled) {
+      if (enabled) {
+        ATH.TELEMETRY_ENABLED = true;
+      } else {
+        ATH.TELEMETRY_ENABLED = false;
+      }
+    }.bind(this);
+
+    SettingsListener.observe(ATH.TELEMETRY_ENABLED_KEY,
+      false, this.telemetryEnabledListener);
+    ATH.SETTINGS_LISTENER = true;
+  };
+
+  // API for count histogram entry.
+  // <name> is the unique name of the metric.
+  // This function increments the count for a histogram of type count.
+  AdvancedTelemetryHelper.prototype.count = function(name) {
+    if (!ATH.TELEMETRY_ENABLED) {
+      return;
+    }
+
+    var message;
+ 
+    if (name) {
+      message = APP_TELEMETRY_LOG_PREFIX;
+
+      message += '|' + name;
+      // For a counter histogram, the value makes no difference, just adding
+      // 1 here to keep the parsing consistent.  The last parameter is the
+      // constant Histogram type -- in this case HISTOGRAM_COUNT = 4
+      message += '|1|4';
+      console.info(message);
+    }
+    else {
+      console.warn('A counter telemetry entry requires one parameter');
+    }
+  };
+
+  // API for a linear histogram entry.
+  // <name>: Unique name of the metric
+  // <value>: Numeric value for this histogram.  Should fall between max/min.
+  // <min>: the minimum numeric value for this histogram.
+  // <max>: the maximum numeric value for this histogram
+  // <buckets>: the total number of buckets for this histogram.
+  // min/max/buckets are required on the first call.  For subsequent calls they
+  // are optional.
+  AdvancedTelemetryHelper.prototype.linear =
+  function(name, value, min, max, buckets) {
+    if (!ATH.TELEMETRY_ENABLED) {
+      return;
+    }
+
+    var message;
+
+    if (name && value) {
+      message = APP_TELEMETRY_LOG_PREFIX;
+
+      message += '|' + name;
+      message += '|' + value + '|1|';
+      if(typeof min !== 'undefined' && typeof max !== 'undefined' &&
+                                       typeof buckets !== 'undefined') {
+        message += min + '|' + max + '|' + buckets;
+      }
+      console.info(message);
+    }
+    else {
+      console.warn('A telemetry log entry requires two or five parameters');
+    }
+  };
+
```

Comment:
The 'exp' function seems identical to the 'linear' function expect for the histogram type. Perhaps it's worth refactoring to remove the duplicate code.